### ISSUE TITLE
actor: Animations not working until actor defined as part of class

### ISSUE THREAD

**Issue by Benjamin8693** [panda3d/panda3d#922]
Today I was doing some experimentation with Actors, when I ran into quite a weird issue. Originally, my code looked something like this:

```
model = Actor.Actor('my-mod.bam', {'spin': 'my-spin.bam'})
model.loop('spin')
```

For some reason, my actor was not doing his animations. I tried swapping out actors to make sure the one I was using didn't have something faulty with the model, but the issue kept happening. I asked a friend about this, and he told me he had this issue before. He recommended that I define my actor as part of the class, so I did. My code then looked like this:

```
self.model = Actor.Actor('my-mod.bam', {'spin': 'my-spin.bam'})
self.model.loop('spin')
```

Only after I did this was my actor animating correctly. I was a bit confused when it began to work, so I decided to post an issue here about it.

Thanks,
Benjamin

**Comment by Moguri** [comment]
This is a known quirk of the `Actor` class. However, I was surprised to not find an existing issue for it.

When an Actor instance is destroyed, it will stop playing animations (I don't remember if the animation is explicitly stopped, or just no longer updating), and you need to keep a reference around to avoid having the instance garbage collected. This is why adding it as a variable to another instance solves the problem (the ref count is incremented and thus doesn't go to zero).

I agree this is strange behavior and not very newbie friendly. If we can fix this in a non-API breaking way I would be all for it. However, it would break games that were relying on this (i.e., expecting animations to stop when the instance went out of scope).

**Comment by rdb** [comment]
Fundamentally, the problem is with Actor's design.  It is a Python class, therefore not living in the scene graph, extending a C++ class that does live in the scene graph.  So Panda continues to keep a reference to the C++ bit (the unanimated model) even when the Python bit (containing the animations) goes out of scope.

I think we probably want to replace Actor with a C++ interface.  That might involve improving the interface of Character to support all the things Actor does.  Or, creating a new C++-based interface to manage the AnimControls playing on a node.
